lib/code_corps/donation_goals_manager.ex

```diff
@@ -0,0 +1,114 @@
+defmodule CodeCorps.DonationGoalsManager do
+  import Ecto.Query
+
+  alias CodeCorps.{DonationGoal, Project, Repo}
+  alias Ecto.Multi
+
+  def create(attributes) do
+    changeset = %DonationGoal{} |> DonationGoal.create_changeset(attributes)
+
+    multi = Multi.new
+    |> Multi.insert(:donation_goal, changeset)
+    |> Multi.run(:update_current_goal, &transactional_update_current_goal/1)
+
+    case Repo.transaction(multi) do
+      {:ok, %{donation_goal: donation_goal, update_current_goal: _}} ->
+        {:ok, Repo.get(DonationGoal, donation_goal.id)}
+      {:error, :donation_goal, %Ecto.Changeset{} = changeset, %{}} ->
+        {:error, changeset}
+      {:error, _failed_operation, _failed_value, _changes_so_far} ->
+        {:error, :unhandled}
+    end
+  end
+
+  def update(donation_goal, attributes) do
+    changeset = donation_goal |> DonationGoal.create_changeset(attributes)
+
+    multi = Multi.new
+    |> Multi.update(:donation_goal, changeset)
+    |> Multi.run(:update_current_goal, &transactional_update_current_goal/1)
+
+    case Repo.transaction(multi) do
+      {:ok, %{donation_goal: donation_goal, update_current_goal: _}} ->
+        {:ok, Repo.get(DonationGoal, donation_goal.id)}
+      {:error, :donation_goal, %Ecto.Changeset{} = changeset, %{}} ->
+        {:error, changeset}
+      {:error, _failed_operation, _failed_value, _changes_so_far} ->
+        {:error, :unhandled}
+    end
+  end
+
+  def update_current_goal(%Project{} = project) do
+    unset_all_goals_from_current(project)
+    set_new_current_goal(project)
+  end
+
+  defp transactional_update_current_goal(%{donation_goal: %DonationGoal{project_id: project_id}}) do
```

Comment:
1. The name of this is misleading. It implies that it will wrap its behavior in a transaction, which it doesn't. Instead, it expects to be used _inside_ a transaction.
2. I don't believe that you need a `Project` record at this point, therefore the function itself seems like its doing extra work.